app/services/translate_status_service.rb

```diff
@@ -32,6 +40,69 @@ def languages
   end
 
   def content_hash
-    Digest::SHA256.base64digest(@content)
+    Digest::SHA256.base64digest(@source_texts.transform_keys { |key| key.respond_to?(:id) ? "#{key.class}-#{key.id}" : key }.to_json)
+  end
+
+  def source_texts
+    texts = {}
+    texts[:content] = prerender_custom_emojis(status_content_format(@status)) if @status.content.present?
+    texts[:spoiler_text] = prerender_custom_emojis(@status.spoiler_text) if @status.spoiler_text.present?
+
+    @status.preloadable_poll&.loaded_options&.each do |option|
+      texts[option] = prerender_custom_emojis(option.title)
+    end
+
+    @status.media_attachments.each do |media_attachment|
+      texts[media_attachment] = media_attachment.description
+    end
+
+    texts
+  end
+
+  def build_status_translation(translations)
+    status_translation = StatusTranslation.new(
+      detected_source_language: translations.first&.detected_source_language,
+      language: @target_language,
+      provider: translations.first&.provider,
+      content: '',
+      spoiler_text: '',
+      poll_options: [],
+      media_attachments: []
+    )
+
+    @source_texts.keys.each_with_index do |source, index|
+      translation = translations[index]
+
+      case source
+      when :content
+        status_translation.content = detect_custom_emojis(translation.text).to_html
+      when :spoiler_text
+        status_translation.spoiler_text = detect_custom_emojis(translation.text).content
+      when Poll::Option
+        status_translation.poll_options << StatusTranslation::Option.new(
+          title: detect_custom_emojis(translation.text).content
+        )
+      when MediaAttachment
+        status_translation.media_attachments << StatusTranslation::MediaAttachment.new(
+          id: source.id,
+          description: translation.text
+        )
+      end
+    end
+
+    status_translation
+  end
+
+  def prerender_custom_emojis(text)
+    EmojiFormatter.new(html_escape(text), @status.emojis, { data_shortcode: true }).to_s
```

Comment:
We don't pre-render custom emoji as image tags in the API, so this is a big change. Image tags are not expected in the returned HTML. I would just wrap shortcodes in `<span class="notranslate">`

Reply:
Good point.

I have verified that `<span translate="no">:tag:</span>` works with both DeepL and LibreTranslate (the standardized [`translate` attribute](https://developer.mozilla.org/en-US/docs/Web/HTML/Global_attributes/translate) is probably more reliable than `class`). So I agree that that is a simpler+better approach. I'll look into that.